src/CloudDataProvider.ts

```diff
@@ -1,83 +1,92 @@
   kind: "cloudApp";
+  domain: string;
   app: DbosCloudApp;
-  credentials: DbosCloudCredentials;
 };
 
-export interface CloudDatabaseNode {
-  kind: "cloudDatabase";
-  database: DbosCloudDatabase;
-  credentials: DbosCloudCredentials;
+export interface CloudDbInstanceNode {
+  kind: "cloudDbInstance";
+  domain: string;
+  dbInstance: DbosCloudDbInstance;
 }
 
-type CloudProviderNode = CloudDomainNode | CloudServiceTypeNode | CloudAppNode | CloudDatabaseNode;
+type CloudProviderNode = CloudDomainNode | CloudResourceTypeNode | CloudAppNode | CloudDbInstanceNode;
 
 export class CloudDataProvider implements vscode.TreeDataProvider<CloudProviderNode> {
   private readonly onDidChangeTreeDataEmitter = new vscode.EventEmitter<CloudProviderNode | CloudProviderNode[] | undefined | null | void>();
   readonly onDidChangeTreeData = this.onDidChangeTreeDataEmitter.event;
 
-  private readonly domains = new Set<string>();
+  private readonly domains: Array<CloudDomainNode>;
+  private readonly apps = new Map<string, CloudAppNode[]>();
+  private readonly dbInstances = new Map<string, CloudDbInstanceNode[]>();
 
   constructor() {
     const { cloudDomain } = getCloudDomain();
-    this.domains.add(cloudDomain);
+    this.domains = [{ kind: "cloudDomain", domain: cloudDomain }];
   }
 
-  async #getCredentials(domain?: string | DbosCloudDomain): Promise<DbosCloudCredentials | undefined> {
-    const storedCredentials = await config.getStoredCloudCredentials(domain);
-    if (storedCredentials) {
-      return storedCredentials;
+  async refresh(domain: string) {
+    this.apps.delete(domain);
+    this.dbInstances.delete(domain);
+
+    const node = this.domains.find(d => d.domain === domain);
+    if (node) {
+      this.onDidChangeTreeDataEmitter.fire(node);
     }
-    return await config.cloudLogin(domain);
   }
 
   async getChildren(element?: CloudProviderNode | undefined): Promise<CloudProviderNode[]> {
     if (element === undefined) {
-      const children = new Array<CloudDomainNode>();
-      for (const domain of this.domains) {
-        const credentials = await config.getStoredCloudCredentials(domain);
-        children.push({ kind: 'cloudDomain', domain, credentials });
-      }
-      return children;
+      return this.domains;
     }
 
     if (element.kind === "cloudDomain") {
-      return [
-        <CloudServiceTypeNode>{ kind: 'cloudServiceType', domain: element.domain, credentials: element.credentials, type: "Applications" },
-        <CloudServiceTypeNode>{ kind: 'cloudServiceType', domain: element.domain, credentials: element.credentials, type: "Databases" },
-      ];
+      if (!this.apps.has(element.domain) || !this.dbInstances.has(element.domain)) {
+        const credentials = await config.getCredentials(element.domain);
+        if (!credentials) { return []; }
+        const [apps, dbInstances] = await Promise.all([listApps(credentials), listDbInstances(credentials)]);
+        if (isUnauthorized(apps)) {
```

Comment:
Can we prompt some error messages if we get "unauthorized" from DBOS Cloud?

Reply:
Adding a call to validateCredentials above which triggers login flow